File: system-test/java/src/main/java/com/mycompany/myshop/testkit/dsl/core/usecase/external/clock/usecases/ReturnsTime.java
package com.mycompany.myshop.testkit.dsl.core.usecase.external.clock.usecases;

import com.mycompany.myshop.testkit.driver.port.external.clock.ClockDriver;
import com.mycompany.myshop.testkit.driver.port.external.clock.dtos.ReturnsTimeRequest;
import com.mycompany.myshop.testkit.dsl.core.usecase.external.clock.usecases.base.BaseClockUseCase;
import com.mycompany.myshop.testkit.dsl.core.shared.UseCaseResult;
import com.mycompany.myshop.testkit.dsl.core.shared.UseCaseContext;
import com.mycompany.myshop.testkit.dsl.core.shared.VoidVerification;
import com.mycompany.myshop.testkit.driver.port.dtos.error.SystemError;

public class ReturnsTime extends BaseClockUseCase<Void, VoidVerification> {
    private String time;

    public ReturnsTime(ClockDriver driver, UseCaseContext context) {
        super(driver, context);
    }

    public ReturnsTime time(String time) {
        this.time = time;
        return this;
    }

    @Override
    public UseCaseResult<Void, VoidVerification> execute() {
        var request = ReturnsTimeRequest.builder()
                .time(time)
                .build();

        var result = driver.returnsTime(request);

        return new UseCaseResult<>(result.mapError(e -> SystemError.of(e.getMessage())), context, VoidVerification::new);
    }
}

File: system-test/typescript/src/testkit/dsl/core/usecase/external/clock/usecases/ReturnsTime.ts
import type { ClockDriver } from '../../../../../../driver/port/external/clock/clock-driver.js';
import type { ReturnsTimeRequest } from '../../../../../../driver/port/external/clock/dtos/ReturnsTimeRequest.js';
import { success, failure, type Result } from '../../../../../../common/result.js';
import type { SystemError } from '../../../../../../driver/port/dtos/errors/SystemError.js';
import { UseCaseResult } from '../../../../shared/use-case-result.js';
import { VoidVerification } from '../../../../shared/void-verification.js';
import type { UseCaseContext } from '../../../../shared/use-case-context.js';
import { BaseClockUseCase } from './base/BaseClockUseCase.js';

export class ReturnsTime extends BaseClockUseCase<void, VoidVerification> {
  private _time: string = '';

  constructor(driver: ClockDriver, context: UseCaseContext) {
    super(driver, context);
  }

  time(time: string): this {
    this._time = time;
    return this;
  }

  async execute(): Promise<UseCaseResult<void, VoidVerification>> {
    const request: ReturnsTimeRequest = {
      time: this._time,
    };
    const driverResult = await this.driver.returnsTime(request);
    const result: Result<void, SystemError> = driverResult.success
      ? success(undefined)
      : failure({ message: driverResult.error.message ?? 'Clock error', fieldErrors: [] });

    return new UseCaseResult(
      result,
      this.context,
      (_, ctx) => new VoidVerification(undefined, ctx),
    );
  }
}
